File: backend/schemas/analytics.py
"""
Pydantic models for analytics data structures.
"""
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel


class Metric(BaseModel):
    """A data metric."""
    name: str
    value: float
    unit: Optional[str] = None
    timestamp: datetime


class DataPoint(BaseModel):
    """A single data point in a time series."""
    timestamp: datetime
    value: float
    labels: Optional[Dict[str, str]] = None


class TimeSeries(BaseModel):
    """A time series dataset."""
    name: str
    data_points: List[DataPoint]
    metrics: Optional[Dict[str, float]] = None


class Trend(BaseModel):
    """A trend analysis result."""
    direction: str  # "increasing", "decreasing", "stable"
    magnitude: float
    confidence: float
    start_date: datetime
    end_date: datetime
    description: str


class Anomaly(BaseModel):
    """An anomaly detection result."""
    timestamp: datetime
    value: float
    expected_range: tuple
    severity: str  # "low", "medium", "high"
    description: str


class Correlation(BaseModel):
    """A correlation analysis result."""
    metric1: str
    metric2: str
    coefficient: float
    significance: float
    description: str


class AnalyticsResult(BaseModel):
    """Result of an analytics operation."""
    type: str
    data: Dict[str, Any]
    summary: str
    confidence: float
    timestamp: datetime

File: frontend/pbiviz/src/types/analytics.ts
/**
 * Analytics types and interfaces
 */

export interface Metric {
  name: string;
  value: number;
  unit?: string;
  timestamp: string;
}

export interface DataPoint {
  timestamp: string;
  value: number;
  labels?: Record<string, string>;
}

export interface TimeSeries {
  name: string;
  dataPoints: DataPoint[];
  metrics?: Record<string, number>;
}

export interface Trend {
  direction: 'increasing' | 'decreasing' | 'stable';
  magnitude: number;
  confidence: number;
  startDate: string;
  endDate: string;
  description: string;
}

export interface Anomaly {
  timestamp: string;
  value: number;
  expectedRange: [number, number];
  severity: 'low' | 'medium' | 'high';
  description: string;
}

export interface Correlation {
  metric1: string;
  metric2: string;
  coefficient: number;
  significance: number;
  description: string;
}

export interface AnalyticsResult {
  type: string;
  data: Record<string, any>;
  summary: string;
  confidence: number;
  timestamp: string;
}
